README.md
# Quick Start
This is a standard expressjs server. By defaul it serves on all interfaces and on port 3000.

```shell
npm i express-vanilla
```

Create an index.js file:

```javascript
const initialiseServer = require('express-vanilla')

const routes = [
  middleware1,
  middleware2,
  routes1,
  routes2,
]

const server = initialiseServer(routes)

```

# Details

It assumes that static content resolves to the ./public folder. i.e. http://hostname:port/index.html would resolve to ./index.html

You can configure the interface and port with the following env vars:

```console
PORT=3001
HOST=10.19.0.1
```

# Using this server with TLS

Google and Chrome are beginning to score non-TLS sites lower and warning of insecure access. This is primarily why the default port is not 80. The choice of port 3000 is informed by Docker microservices where one or two ports exposed is sufficient and 3000 is usually the service endpoint.

I suggest implementing a reverse proxy to provide SSL termination. Nginx works well, is easy to set up and maintain, and is performant. It is suitable for most cases where traffic is reasonable (don't know exact figures - I would expect a setup to cope fine for 300 requests per minute). Any higher loads should be possible to resolve with a load balancer and additional servers. 

The middleware and routers will be applied to the server in the order they have been added into the array. 

The initialiser returns the express listener which you can attached event handlers to and request a graceful shutdown, etc.

# Default request handlers

It implements an 404 handler for routes which aren't found. If you are creating an SPA, you may prefer to return a default page by appending a router which 
lists .get('*', {function}) as one of its routers.

It implements a standard error handler. You can override it by adding an alternative in the last array element


RELEASE_NOTES.md
# V0.0.8 Add default static file for / route
- Support passing a single request handler as an alternative to an array of them.
- add a default route to return index.html
- add a test script

# V0.0.1 First release of a standard expressjs configuration
- Defaults to dev (cors allowed, port 3000 on all interfaces, serves static files from ./public, 'dev' config for morgan)
- Uses helmet for security
- Simple error handling



test/index.js
const initialiseServer = require('../index')

initialiseServer()

index.js
const cors = require('cors')
const path = require('path')
const helmet = require('helmet')
const morgan = require('morgan')
const express = require('express')
const logger = require('./utils/logger')

const isObject = function(handlers) {
  return !!handlers && handlers.constructor === Object
}

const isArray = function(handlers) {
  return !!handlers && handlers.constructor === Array
}

/**
 * Initialise an express js server with helmet implemented for security, morgan for logging serving
 * static content from public and accepting an array of express handlers such as middleware, 
 * request handlers and error handlers.
 * 
 * Assumes a default port of 3000 mounted to all the machines interfaces. See params for customisation.
 * 
 * In development (set NODE_ENV to 'development') it allows all CORS requests. I don't recommend this for prod
 * 
 * @param {Object || Array} arg1 - optional set of otpions. Omit if you want to use the defaults
 * @param {Array || null} arg2 - optional array of handlers such as middleware and ```Router```s (const Router = require('express).Router)
 * @param {Array} options list of customisation options:
 * @param {String} options.port port to mount expressjs to. This can also be provided as an ENV variable PORT. Default 3000
 * @param {String} options.host the interface to mount express to. You can provide a unix socket path for this. Default 0.0.0.0. Can be provided via ENV var HOST_ADDR (or use machine's HOST)
 * @param {Object} options.morganOpts morgan options (https://www.npmjs.com/package/morgan) Default 'dev'
 */
const initialiseServer = async (arg1, arg2) => {
  /////////////////////////
  // validate configuration
  /////////////////////////

  let options
  let suppliedHandlers // routers (request handlers) and middleware

  if(Array.isArray(arg1)) {
    suppliedHandlers = arg1
  } else {
    options = arg1
    suppliedHandlers = arg2
  }

  const port = (options && options.port) ? options.port : process.env.PORT || '3000'
  const host = (options && options.host) ? options.host : process.env.HOST_ADDR || process.env.HOST || '0.0.0.0'
  const morganOpts = (options && options.morganOpts) ? options.morganOpts : 'dev'

  // accept ```handlers``` as object or array, convert ```handlers``` to array if required
  const handlers = isArray(suppliedHandlers) ? suppliedHandlers : isObject(suppliedHandlers) ? [suppliedHandlers] : []

  ///////////////////
  // Configure server
  ///////////////////
  const app = express()

  app.set('port', port)
  app.set('host', host)

  if(process.env.NODE_ENV==='development') {
    app.use(cors())
  }

  app.use(helmet())
  app.use(morgan(morganOpts))

  ///////////////////
  // Serve statics
  ///////////////////
  app.use(express.static(path.join(__dirname, 'public')))

  //////////////////////////////////////////////
  // configure handlers
  // - assumes array of handlers is 
  // - 1) Middleware
  // followed by
  // - 2) request handlers or routers
  // - 3) override default/no router found handler
  // - 4) override error handler
  //////////////////////////////////////////////
  handlers.forEach(handler => {
    app.use(handler)    
  });

  // viewed at http://localhost:8080
  app.get('/', function(req, res) {
    res.sendFile('/public/index.html');
  });

//////////////////////////////////////////////
  // Catch 404 and forward to error handler
  //////////////////////////////////////////////
  app.use((req, res, next) => {
    const err = new Error(`${new Date()} Route not Found ${req.path}`)
    err.status = 404
    next(err)
  })
  
  //////////////////////////////////////////////
  // Error handler
  //////////////////////////////////////////////
  app.use((error, req, res) => { // eslint-disable-line no-unused-vars
    res
      .status(error.status || 500)
      .send({error})
  })  

  const server = app.listen(app.get('port'), app.get('host'), () => {
    logger.log('info', `Login manager started at http://${app.get('host')}:${app.get('port')}`)
  })

  return server
}

module.exports = initialiseServer
